zuul-core/src/main/java/com/netflix/zuul/netty/filter/BaseZuulFilterRunner.java

```diff
@@ -474,7 +475,12 @@ public void onError(Throwable ex) {
         public void onCompleted() {
             try (TaskCloseable ignored = PerfMark.traceTask(filter, f -> f.filterName() + ".onCompletedAsync")) {
                 PerfMark.linkIn(onCompletedLinkOut.get());
+                if (outMesg == null) {
+                    outMesg = filter.getDefaultOutput(inMesg);
+                }
                 decrementConcurrency();
+                recordFilterCompletion(ExecutionStatus.SUCCESS, filter, startTime, inMesg, snapshot);
+                resumeInBindingContext(outMesg, filter.filterName());
```

Comment:
A nice side-effect of this may be `onCompleted()` actually guarantees exactly once semantics.